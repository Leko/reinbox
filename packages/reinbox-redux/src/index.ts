export { actions } from "./actions";
export { createReducer } from "./createReducer";
export { default as Provider } from "./ProviderContainer";
