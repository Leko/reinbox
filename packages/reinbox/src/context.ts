import * as React from "react";

export type Payload = {
  id: string;
} & {
  [otherPropName: string]: any;
};

export type Topic = {
  [payloadId: string]: Payload;
};

export type Topics = {
  [topicName: string]: Topic;
};

export type State = {
  topics: Topics;
};

export type Context = {
  state: State;
  publish: (topic: string, payload: Payload) => void;
  dismiss: (topic: string, id: string) => void;
};

const defaultContext: Context = {
  state: {
    topics: {}
  },
  publish: () => {},
  dismiss: () => {}
};

const { Provider, Consumer } = React.createContext(defaultContext);

export { Provider, Consumer };
