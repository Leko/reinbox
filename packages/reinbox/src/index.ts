export { InboxProvider as Provider } from "./InboxProvider";
export { Inbox } from "./Inbox";
export { withPublish, WithPublishProps } from "./withPublish";
