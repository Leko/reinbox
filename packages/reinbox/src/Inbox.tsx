import * as React from "react";
import { Consumer, Payload } from "./context";

type Props = {
  topic: string;
  children: (
    context: { payload: Payload | null; dismiss: () => void }
  ) => React.ReactNode;
};

export const Inbox = ({ topic, children }: Props) => (
  <Consumer>
    {({ state: { topics }, dismiss }) => {
      const payloads = topics[topic] || {};
      const payload = Object.values(payloads)[0] || null;
      return children({
        payload,
        dismiss: () => (payload ? dismiss(topic, payload.id) : null)
      });
    }}
  </Consumer>
);
