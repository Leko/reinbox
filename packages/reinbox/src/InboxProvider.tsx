import * as React from "react";
// @ts-ignore power-assign not provided DTS
import { assign } from "power-assign";
import { Provider, State, Topics, Payload } from "./context";

type Props = {
  initialTopics?: Topics;
  children: React.ReactNode;
};

export class InboxProvider extends React.Component<Props, State> {
  state: State = {
    topics: this.props.initialTopics || {}
  };

  publish = (topic: string, payload: Payload) => {
    const newState = assign(this.state, {
      [`topics.${topic}.${payload.id}`]: payload
    });
    this.setState(newState);
  };

  dismiss = (topic: string, id: string) => {
    const newState = assign(this.state, {
      $unset: {
        [`topics.${topic}.${id}`]: ""
      }
    });
    this.setState(newState);
  };

  render() {
    const { state, dismiss, publish } = this;
    const { children } = this.props;

    return (
      <Provider
        value={{
          state,
          dismiss,
          publish
        }}
      >
        {children}
      </Provider>
    );
  }
}
