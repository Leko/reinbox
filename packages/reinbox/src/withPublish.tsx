import * as React from "react";
import { Consumer, Payload } from "./context";

export type WithPublishProps = {
  publish: (topic: string, payload: Payload) => void;
};

export const withPublish = <P extends {}>(
  Cmp: React.ComponentType<P & WithPublishProps>
): React.ComponentType<P> => (props: P) => (
  <Consumer>
    {({ publish }) => (
      <Cmp
        publish={(topic: string, payload: Payload) => publish(topic, payload)}
        {...props}
      />
    )}
  </Consumer>
);
