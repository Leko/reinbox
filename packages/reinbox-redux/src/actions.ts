import { Payload } from "reinbox";

type PublishAction = {
  type: "reinbox/publish";
  payload: {
    topic: string;
    payload: Payload;
  };
};
type DismissAction = {
  type: "reinbox/dismiss";
  payload: {
    topic: string;
    id: string;
  };
};
export type Actions = PublishAction | DismissAction;

const publish = (topic: string, payload: Payload): PublishAction => ({
  type: "reinbox/publish",
  payload: {
    topic,
    payload
  }
});
const dismiss = (topic: string, id: string): DismissAction => ({
  type: "reinbox/dismiss",
  payload: {
    topic,
    id
  }
});

export const actions = {
  publish,
  dismiss
};
