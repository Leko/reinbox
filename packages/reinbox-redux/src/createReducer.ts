// @ts-ignore power-assign not provided DTS
import { assign } from "power-assign";
import { State } from "reinbox";
import { Actions } from "./actions";

export const createReducer = () => {
  const initialState: State = {
    topics: {}
  };

  const reducer = (state: State = initialState, action: Actions) => {
    switch (action.type) {
      case "reinbox/publish": {
        const { topic, payload } = action.payload;
        return assign(state, {
          [`topics.${topic}.${payload.id}`]: payload
        });
      }
      case "reinbox/dismiss": {
        const { topic, id } = action.payload;
        return assign(state, {
          $unset: {
            [`topics.${topic}.${id}`]: ""
          }
        });
      }
      default:
        return state;
    }
  };

  return reducer;
};
