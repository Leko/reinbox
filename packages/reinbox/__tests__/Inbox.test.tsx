import * as React from "react";
import * as renderer from "react-test-renderer";
import { Inbox } from "../src/Inbox";
import { InboxProvider } from "../src/InboxProvider";

test("snapshot for empty children", () => {
  const component = renderer.create(<Inbox topic="test">{() => null}</Inbox>);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("snapshot for static children", () => {
  const component = renderer.create(
    <Inbox topic="test">
      {({ dismiss }) => <button onClick={dismiss}>click me</button>}
    </Inbox>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("snapshot for static children with payload", () => {
  const topic = "test";
  const component = renderer.create(
    <InboxProvider
      initialTopics={{
        [topic]: { xxx: { id: "xxx", message: "notification message" } }
      }}
    >
      <Inbox topic={topic}>
        {({ payload, dismiss }) =>
          payload && (
            <div>
              <span>{payload.message}</span>
              <button onClick={dismiss}>dismiss</button>
            </div>
          )
        }
      </Inbox>
    </InboxProvider>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("Inbox should show first of message", () => {
  const topic = "test";
  const messages = {
    xxx: { id: "xxx" },
    yyy: { id: "yyy" },
    zzz: { id: "zzz" }
  };
  const component = renderer.create(
    <InboxProvider
      initialTopics={{
        [topic]: messages
      }}
    >
      <Inbox topic={topic}>
        {({ payload }) => (payload ? payload.id : null)}
      </Inbox>
    </InboxProvider>
  );
  let tree = component.toJSON();
  expect(tree).toBe("xxx");
});

test("It should show next message when called dismiss", () => {
  const topic = "test";
  const messages = {
    xxx: { id: "xxx" },
    yyy: { id: "yyy" }
  };
  const component = renderer.create(
    <InboxProvider
      initialTopics={{
        [topic]: messages
      }}
    >
      <Inbox topic={topic}>
        {({ payload, dismiss }) => (
          <button onClick={dismiss}>{payload ? payload.id : null}</button>
        )}
      </Inbox>
    </InboxProvider>
  );
  let tree = component.toJSON();
  // @ts-ignore
  tree.props.onClick();
  tree = component.toJSON();
  // @ts-ignore
  expect(tree.children[0]).toBe("yyy");
});

test("It should pass null with single message when called dismiss", () => {
  const topic = "test";
  const messages = {
    xxx: { id: "xxx" }
  };
  const component = renderer.create(
    <InboxProvider
      initialTopics={{
        [topic]: messages
      }}
    >
      <Inbox topic={topic}>
        {({ payload, dismiss }) => (
          <button onClick={dismiss}>{payload ? payload.id : null}</button>
        )}
      </Inbox>
    </InboxProvider>
  );
  let tree = component.toJSON();
  // @ts-ignore
  tree.props.onClick();
  tree = component.toJSON();
  // @ts-ignore
  expect(tree.children).toBe(null);
});
