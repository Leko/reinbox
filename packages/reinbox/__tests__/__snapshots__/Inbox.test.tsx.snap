// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot for empty children 1`] = `null`;

exports[`snapshot for static children 1`] = `
<button
  onClick={[Function]}
>
  click me
</button>
`;

exports[`snapshot for static children with payload 1`] = `
<div>
  <span>
    notification message
  </span>
  <button
    onClick={[Function]}
  >
    dismiss
  </button>
</div>
`;
